File: test_your_might.go
package main

import (
	"fmt"
	"math"
	"math/rand/v2"
	"runtime"
	"strconv"
	"sync"
	"time"
)

const (
	BET_MULTIPLIER_SCALE uint8  = 100
	TECHNICAL_SYMBOL_ID  uint8  = math.MaxUint8
	TOTAL_ITERATIONS     uint32 = 100000000 // 100M total iterations for Monte Carlo simulation
)

func main() {
	// Static game configurations (3x3 grid base model)
	reelStrips := [][]uint8{
		{4, 0, 1, 3, 2, 5, 5, 1, 4, 0, 2, 3, 5, 1, 1, 0, 4, 2, 3, 5, 0, 4, 1, 2, 3, 5, 0, 4, 2, 1, 3, 5, 0, 2, 4, 1, 3, 5, 0, 2, 1, 4, 3, 5, 0, 1, 2, 4, 3, 5, 0, 1, 4, 2, 3, 5, 0, 4, 1, 2, 3, 5, 0, 4, 2, 1, 3, 5, 0, 2, 4, 1, 3, 5, 0, 2, 1, 4, 3, 5, 0, 1, 2, 4, 3, 5, 0, 1, 4, 2, 3, 5, 0, 4, 1, 2, 3, 5, 0, 4},
		{1, 5, 2, 0, 3, 4, 4, 2, 1, 5, 3, 0, 4, 2, 2, 5, 1, 3, 0, 4, 5, 1, 2, 3, 0, 4, 5, 1, 3, 2, 0, 4, 5, 3, 1, 2, 0, 4, 5, 3, 2, 1, 0, 4, 5, 2, 3, 1, 0, 4, 5, 2, 1, 3, 0, 4, 5, 1, 2, 3, 0, 4, 5, 1, 3, 2, 0, 4, 5, 3, 1, 2, 0, 4, 5, 3, 2, 1, 0, 4, 5, 2, 3, 1, 0, 4, 5, 2, 1, 3, 0, 4, 5, 1, 2, 3, 0, 4, 5, 1},
		{3, 2, 0, 4, 5, 1, 1, 0, 3, 2, 5, 4, 1, 0, 0, 2, 3, 5, 4, 1, 2, 3, 0, 5, 4, 1, 2, 3, 5, 0, 4, 1, 2, 5, 3, 0, 4, 1, 2, 5, 0, 3, 4, 1, 2, 0, 5, 3, 4, 1, 2, 0, 3, 5, 4, 1, 2, 3, 0, 5, 4, 1, 2, 3, 5, 0, 4, 1, 2, 5, 3, 0, 4, 1, 2, 5, 0, 3, 4, 1, 2, 0, 5, 3, 4, 1, 2, 0, 3, 5, 4, 1, 2, 3, 0, 5, 4, 1, 2, 3},
	}
	winLines := [][]uint8{{0, 1, 2}, {3, 4, 5}, {6, 7, 8}}
	gameFieldReelsHeights := []uint8{3, 3, 3}

	// Utilize available physical threads
	cores := runtime.NumCPU()
	var betCoins uint8 = 1

	// Precomputed and pre-allocated metadata to eliminate runtime overhead
	reelStripsLengths := []uint32{
		uint32(len(reelStrips[0])),
		uint32(len(reelStrips[1])),
		uint32(len(reelStrips[2])),
	}

	payoutConfig := map[string]uint16{
		"AAA": 10, "WAA": 10, "AWA": 10, "WWA": 10, "AAW": 10, "WAW": 10, "AWW": 10, "WWW": 30,
		"BBB": 50, "WBB": 50, "BWB": 50, "WWB": 50, "BBW": 50, "WBW": 50, "BWW": 50,
		"CCC": 100, "WCC": 100, "CWC": 100, "WWC": 100, "CCW": 100, "WCW": 100, "CWW": 100,
	}

	symbolIDsByName := map[byte]uint8{'S': 0, 'A': 1, 'B': 2, 'C': 3, 'W': 4, 'D': 5}
	symbolCount := len(symbolIDsByName) // Total number of unique symbols (e.g., 6)

	// 3D Look-Up Table (LUT) initialized during engine bootstrapping
	PayOutLookUpTable := make([][][]uint16, symbolCount)
	for i := 0; i < symbolCount; i++ {
		PayOutLookUpTable[i] = make([][]uint16, symbolCount)
		for j := 0; j < symbolCount; j++ {
			PayOutLookUpTable[i][j] = make([]uint16, symbolCount)
		}
	}

	for combinationKey, betMultiplier := range payoutConfig {
		res1 := symbolIDsByName[combinationKey[0]]
		res2 := symbolIDsByName[combinationKey[1]]
		res3 := symbolIDsByName[combinationKey[2]]

		PayOutLookUpTable[res1][res2][res3] = betMultiplier
	}
	var totalReels uint8 = uint8(len(gameFieldReelsHeights))

	// Async workloads orchestration
	iterationsPerWorker := TOTAL_ITERATIONS / uint32(cores)
	var wg sync.WaitGroup
	winsChannel := make(chan uint64, cores)

	startTime := time.Now()

	// High-performance non-blocking seed generation
	baseServerSeed := uint64(time.Now().UnixNano())
	baseClientSeed := uint64(123456789)

	for coreNumber := 0; coreNumber < cores; coreNumber++ {
		wg.Add(1)

		go func(coreNumber int) {
			defer wg.Done()

			serverSeed := baseServerSeed + uint64(coreNumber)
			clientSeed := baseClientSeed * uint64(coreNumber+1)

			winsChannel <- RunSimulationWorker(
				serverSeed,
				clientSeed,
				iterationsPerWorker,
				reelStrips,
				gameFieldReelsHeights,
				totalReels,
				reelStripsLengths,
				winLines,
				PayOutLookUpTable,
				betCoins,
			)
		}(coreNumber)
	}

	// Wait for processing lanes to complete, then close data pipeline
	wg.Wait()
	close(winsChannel)

	var totalScaledWin uint64
	for win := range winsChannel {
		totalScaledWin += win
	}

	executionTime := time.Since(startTime)
	iterationsPerSecond := float64(TOTAL_ITERATIONS) / executionTime.Seconds()
	rtp := float64(totalScaledWin) / float64(TOTAL_ITERATIONS*uint32(betCoins)) / float64(BET_MULTIPLIER_SCALE)

	// Output Formatting
	fmt.Printf("RTP: %g\n", rtp)
	fmt.Printf("Execution Time: %.4f seconds\n", executionTime.Seconds())
	fmt.Printf("Iterations per second (RPS): %s\n", formatWithSpaces(int(iterationsPerSecond)))
}

// RunSimulationWorker executes pure mathematical simulations on isolated threads.
// All data structures are optimized for Data Locality to ensure L1/L2 Cache hits.
func RunSimulationWorker(
	serverSeed, clientSeed uint64,
	iterationsPerCore uint32,
	reelStrips [][]uint8,
	gameFieldReelsHeights []uint8,
	totalReels uint8,
	reelStripsLengths []uint32,
	winLines [][]uint8,
	payoutLookUpTable [][][]uint16,
	betCoins uint8,
) uint64 {
	// Local thread-safe pseudo-random number generator (v2 PCG)
	localRNG := rand.New(rand.NewPCG(serverSeed, clientSeed))

	var totalGameFieldCells uint8
	for _, height := range gameFieldReelsHeights {
		totalGameFieldCells += uint8(height)
	}

	var localTotalScaledWin uint64

	// Allocation outside the loop guarantees zero heap allocations on the hot path (CPU Stack bound)
	localGeneratedGameField := make([]uint8, totalGameFieldCells)
	for i := range localGeneratedGameField {
		localGeneratedGameField[i] = TECHNICAL_SYMBOL_ID
	}

	// HOT PATH: Avoids dynamic memory operations; safely overrides the preallocated 1D matrix
	for spinCount := 0; spinCount < int(iterationsPerCore); spinCount++ {

		// Step 1: Spin Simulation & Flat Grid Population (Row-Major Order Mapping)
		for reelID := 0; reelID < int(totalReels); reelID++ {
			reelStripLength := gameFieldReelsHeights[reelID]
			// Fixed boundary math to guarantee access to the very last element of the strip
			maximalPositionOnReelStrip := reelStripsLengths[reelID] - uint32(reelStripLength)

			reelStripStartPosition := localRNG.IntN(int(maximalPositionOnReelStrip) + 1)
			reelStrip := reelStrips[reelID]

			for position := 0; position < int(reelStripLength); position++ {
				gameFieldPosition := position*int(totalReels) + reelID
				localGeneratedGameField[gameFieldPosition] = uint8(reelStrip[reelStripStartPosition+position])
			}
		}

		// Step 2: Unrolled Payline Evaluation via Cache-Friendly Array Lookup
		// Loops use explicit length check allowing the Go compiler to perform all (Bounds Check Elimination)
		for winLineID := 0; winLineID < len(winLines); winLineID++ {
			// Static unrolling assuming 3-symbol evaluations for absolute performance
			firstSymbol := localGeneratedGameField[winLines[winLineID][0]]
			secondSymbol := localGeneratedGameField[winLines[winLineID][1]]
			thirdSymbol := localGeneratedGameField[winLines[winLineID][2]]

			payout := payoutLookUpTable[firstSymbol][secondSymbol][thirdSymbol]
			if payout > 0 {
				localTotalScaledWin += uint64(betCoins) * uint64(payout)
			}
		}
	}

	return localTotalScaledWin
}

// Helper function mimicking PHP's number_format space separator for CLI readability
func formatWithSpaces(number int) string {
	stringNumber := strconv.Itoa(number)
	var response []byte
	length := len(stringNumber)
	for i := 0; i < length; i++ {
		if i > 0 && (length-i)%3 == 0 {
			response = append(response, ' ')
		}
		response = append(response, stringNumber[i])
	}

	return string(response)
}

File: test_your_might.php
<?php

declare(strict_types=1);

// CHECKLIST
// - xDebug OFF
// - JIT ON
// - PHP LATEST

// Scale factor for integer-based currency computations to eliminate float precision issues
const MULTIPLIER_SCALE    = 100;
const TECHNICAL_SYMBOL_ID = 255;

// ============================================================================
// DATA LOCALITY & PRECOMPUTATIONS BLOCK (Executed once at startup)
// ============================================================================

// Symbol ID to Character mapping for rapid string-based combo key generation
$gameSymbols = [0 => 'S', 1 => 'A', 2 => 'B', 3 => 'C', 4 => 'W', 5 => 'D'];

// Static reels configuration data (strips mapped as nested arrays)
$reelStrips = [
    [4,0,1,3,2,5,5,1,4,0,2,3,5,1,1,0,4,2,3,5,0,4,1,2,3,5,0,4,2,1,3,5,0,2,4,1,3,5,0,2,1,4,3,5,0,1,2,4,3,5,0,1,4,2,3,5,0,4,1,2,3,5,0,4,2,1,3,5,0,2,4,1,3,5,0,2,1,4,3,5,0,1,2,4,3,5,0,1,4,2,3,5,0,4,1,2,3,5,0,4],
    [1,5,2,0,3,4,4,2,1,5,3,0,4,2,2,5,1,3,0,4,5,1,2,3,0,4,5,1,3,2,0,4,5,3,1,2,0,4,5,3,2,1,0,4,5,2,3,1,0,4,5,2,1,3,0,4,5,1,2,3,0,4,5,1,3,2,0,4,5,3,1,2,0,4,5,3,2,1,0,4,5,2,3,1,0,4,5,2,1,3,0,4,5,1,2,3,0,4,5,1],
    [3,2,0,4,5,1,1,0,3,2,5,4,1,0,0,2,3,5,4,1,2,3,0,5,4,1,2,3,5,0,4,1,2,5,3,0,4,1,2,5,0,3,4,1,2,0,5,3,4,1,2,0,3,5,4,1,2,3,0,5,4,1,2,3,5,0,4,1,2,5,3,0,4,1,2,5,0,3,4,1,2,0,5,3,4,1,2,0,3,5,4,1,2,3,0,5,4,1,2,3]
];

$gameFieldReelsHeights = [3, 3, 3];
$totalReels            = count($gameFieldReelsHeights);

// Pre-allocated flat 1D array representing a 3x3 game field matrix (Data Locality approach)
$generatedGameField = [
    TECHNICAL_SYMBOL_ID, TECHNICAL_SYMBOL_ID, TECHNICAL_SYMBOL_ID,
    TECHNICAL_SYMBOL_ID, TECHNICAL_SYMBOL_ID, TECHNICAL_SYMBOL_ID,
    TECHNICAL_SYMBOL_ID, TECHNICAL_SYMBOL_ID, TECHNICAL_SYMBOL_ID,
];

$betCoins       = 1;
$totalScaledWin = 0;

// Precomputed Win Map Lookup Table (LUT). Converts complex Wild logic checks into O(1) isset operations
$possibleCombinationsPayoutsMap = [
    'AAA' => 10,  'WAA' => 10,  'AWA' => 10,  'WWA' => 10,  'AAW' => 10,  'WAW' => 10,  'AWW' => 10,  'WWW' => 30,
    'BBB' => 50,  'WBB' => 50,  'BWB' => 50,  'WWB' => 50,  'BBW' => 50,  'WBW' => 50,  'BWW' => 50,
    'CCC' => 100, 'WCC' => 100, 'CWC' => 100, 'WWC' => 100, 'CCW' => 100, 'WCW' => 100, 'CWW' => 100,
];

// Flat indices mapping horizontal paylines on the 1D matrix layout
$winLines          = [[0, 1, 2], [3, 4, 5], [6, 7, 8]];
$reelStripsLengths = array_map('count', $reelStrips);
$totalIterations   = 1000000;

// High-precision benchmark entry point measurement
$startTime = microtime(true);

// ============================================================================
// PERFORMANCE HOT PATH: MONTE CARLO EMULATION LOOP
// ============================================================================
for ($spinCount = 1; $spinCount <= $totalIterations; $spinCount++) {

    // Step 1: Spin Simulation & Flat Grid Population
    for ($reelID = 0; $reelID < $totalReels; $reelID++) {
        $reelStrip   = $reelStrips[$reelID];
        $stripLength = $gameFieldReelsHeights[$reelID];

        // Inline RNG window selector preventing index out of bounds exception
        $startPos = mt_rand(0, $reelStripsLengths[$reelID] - (1 + $stripLength));

        // Directly writing into pre-allocated memory layout bypassing intermediate temporary arrays
        for ($position = 0; $position < $stripLength; $position++) {
            $flatIndex                           = $position * $totalReels + $reelID;
            $generatedGameField[(int)$flatIndex] = $reelStrip[$startPos + $position];
        }
    }

    // Step 2: Unrolled Payline Evaluation & Payout Lookup
    foreach ($winLines as $positions) {
        $winComboKey = '';
        foreach ($positions as $pos) {
            $winComboKey .= $gameSymbols[$generatedGameField[$pos]];
        }

        // O(1) hash map evaluation for the generated payout configuration key
        if (isset($possibleCombinationsPayoutsMap[$winComboKey])) {
            $totalScaledWin += $betCoins * $possibleCombinationsPayoutsMap[$winComboKey];
        }
    }
}

// High-precision benchmark exit point measurement
$endTime = microtime(true);

// ============================================================================
// METRICS & METRIC POST-PROCESSING OUTPUT BLOCK
// ============================================================================
$executionTime       = $endTime - $startTime;
$iterationsPerSecond = $executionTime > 0 ? ($totalIterations / $executionTime) : 0;
$rtp                 = $totalScaledWin / ($totalIterations * $betCoins) / MULTIPLIER_SCALE;

echo 'RTP: ' . $rtp . PHP_EOL;
echo 'Execution Time: ' . round($executionTime, 4) . ' seconds' . PHP_EOL;
echo 'Iterations per second (RPS): ' . number_format($iterationsPerSecond, 0, '.', ' ') . PHP_EOL;
